core/messaging/TFMessaging.py

```diff
@@ -0,0 +1,234 @@
+import pandas as pd
+
+from core.const import (
+    INTERNAL_COLUMN_TO_FORM_COLUMN_AND_SECTION,
+    INTERNAL_TABLE_TO_FORM_SHEET,
+    INTERNAL_TYPE_TO_MESSAGE_FORMAT,
+)
+from core.messaging import tf_message_bank as msgs
+from core.messaging.utils import construct_cell_index
+from core.util import get_project_number_by_position, join_as_string
+from core.validation.exceptions import UnimplementedErrorMessageException
+from core.validation.failures.user import (
+    GenericFailure,
+    InvalidEnumValueFailure,
+    NonNullableConstraintFailure,
+    NonUniqueCompositeKeyFailure,
+    UnauthorisedSubmissionFailure,
+    WrongInputFailure,
+    WrongTypeFailure,
+)
+from core.validation.utils import get_cell_indexes_for_outcomes
+
+from ..validation.specific_validations.towns_fund_round_four import (
+    TownsFundRoundFourValidationFailure,
+)
+from . import MessagingBase
+
+
+class TFMessaging(MessagingBase):
+    """Messaging class ABC. Classes that inherit must implement a constructor, and failures_to_message function"""
+
+    def __init__(self, *args):
+        super().__init__(*args)
+        self.tf_msgs = msgs
+
+    def to_message(self, validation_failure):
+        match validation_failure:
+            case NonUniqueCompositeKeyFailure():
+                return self.non_unique_composite_key_failure_message(validation_failure)
+            case WrongTypeFailure():
+                return self.wrong_type_failure_message(validation_failure)
+            case InvalidEnumValueFailure():
+                return self.invalid_enum_value_failure_message(validation_failure)
+            case NonNullableConstraintFailure():
+                return self.non_nullable_constraint_failure_message(validation_failure)
+            case TownsFundRoundFourValidationFailure():
+                return self.towns_fund_round_four_validation_failure(validation_failure)
+            case WrongInputFailure():
+                return self.wrong_input_failure_message(validation_failure)
+            case UnauthorisedSubmissionFailure():
+                return self.unauthorised_submission_failure(validation_failure)
+            case GenericFailure():
+                return self.generic_failure(validation_failure)
+            case _:
+                raise UnimplementedErrorMessageException
+
+    def non_unique_composite_key_failure_message(self, validation_failure) -> tuple[str, str, str, str]:
```

Comment:
We spoke about making these private. Would be good to also update the type hints.